File: subprojects/docs/src/samples/userguide/plugins/buildService/java/buildSrc/src/main/java/DownloadPlugin.java
import org.gradle.api.Project;
import org.gradle.api.Plugin;
import org.gradle.api.provider.Provider;

public class DownloadPlugin implements Plugin<Project> {
    public void apply(Project project) {
        // Register the service
        Provider<WebServer> serviceProvider = project.getGradle().getSharedServices().registerIfAbsent("web", WebServer.class, spec -> {
            spec.getParameters().getPort().set(5005);
        });

        // Connect the provider to the task
        project.getTasks().register("download", Download.java, task -> {
            task.getServer().set(serviceProvider);
        });
    }
}

File: subprojects/docs/src/samples/userguide/plugins/buildService/kotlin/buildSrc/src/main/kotlin/DownloadPlugin.kt
import org.gradle.api.Project
import org.gradle.api.Plugin
import org.gradle.api.provider.Provider

class DownloadPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        // Register the service
        val serviceProvider = project.gradle.sharedServices.registerIfAbsent("web", WebServer::class.java) {
            it.parameters.port.set(5005)
        }

        // Connect the provider to the task
        project.tasks.register("download", Download::class.java) {
            it.server.set(serviceProvider)
            it.outputFile.set(project.layout.buildDirectory.file("result.zip"))
        }
    }
}
